File: BOJ/16XXX/16928.py
"""
https://www.acmicpc.net/problem/16928
16928.뱀과 사다리 게임
실버2
BFS
풀이1. 1336ms
"""

import sys
from collections import deque
N, M = map(int, sys.stdin.readline().split())
jump = dict()
visited = [0] * (101)


def bfs(x):
    q = deque()
    visited[x] = 1
    q.append(x)
    while q:
        x = q.popleft()
        if x == 100:
            return visited[x] - 1
        if x in jump:
            nx = jump[x]
            if visited[nx] and visited[nx] < visited[x]:
                continue
            visited[nx] = visited[x]
            q.appendleft(nx)
        else:
            for nx in range(x+1, x+7):
                if nx > 100 or (visited[nx] and visited[nx] < visited[x] + 1):
                    continue
                visited[nx] = visited[x] + 1
                q.append(nx)


for _ in range(N+M):
    x, y = map(int, sys.stdin.readline().split())
    jump[x] = y

print(bfs(1))

File: kotlin/src/main/kotlin/boj/16000/16928.kt
package boj.`16000`

/*
https://www.acmicpc.net/problem/16928
16928.뱀과 사다리 게임
풀이1.180ms
 */

fun main() {
    val (n, m) = readln().split(" ").map { it.toInt() }
    val jump = mutableMapOf<Int, Int>()
    val visited = BooleanArray(101) { false }

    repeat(n + m) {
        val (from, to) = readln().split(" ").map { it.toInt() }
        jump[from] = to
    }

    val queue = ArrayDeque<Pair<Int, Int>>().apply { add(1 to 0) }
    visited[1] = true

    while (queue.isNotEmpty()) {
        val (x, distance) = queue.removeFirst()

        if (x == 100) {
            println(distance)
            break
        }

        for (dx in 1..6) {
            var nx = x + dx
            if (nx > 100 || visited[nx]) continue
            visited[nx] = true
            queue.add(Pair(nx, distance + 1))

            while(jump.containsKey(nx)) {
                queue.removeLast()
                val nnx = jump[nx]!!
                if (visited[nnx]) break
                visited[nnx] = true
                queue.add(Pair(nnx, distance + 1))
                nx = nnx
            }

        }
    }

}

